sdk/eventhubs/azure-messaging-eventhubs/src/processor.cpp

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "azure/messaging/eventhubs/processor.hpp"
+
+#include "azure/messaging/eventhubs/models/management_models.hpp"
+#include "azure/messaging/eventhubs/models/partition_client_models.hpp"
+#include "private/processor_load_balancer.hpp"
+
+#include <azure/core/diagnostics/logger.hpp>
+#include <azure/core/internal/diagnostics/log.hpp>
+
+#include <iomanip>
+
+using namespace Azure::Core::Diagnostics::_internal;
+using namespace Azure::Core::Diagnostics;
+
+namespace Azure { namespace Messaging { namespace EventHubs {
+
+  Processor::Processor(
+      std::shared_ptr<ConsumerClient> consumerClient,
+      std::shared_ptr<CheckpointStore> checkpointStore,
+      ProcessorOptions const& options)
+      : m_defaultStartPositions(options.StartPositions),
+        m_maximumNumberOfPartitions{options.MaximumNumberOfPartitions},
+        m_checkpointStore(checkpointStore), m_consumerClient(consumerClient),
+        m_prefetch(options.Prefetch), m_nextPartitionClients{}
+  {
+    m_ownershipUpdateInterval = options.UpdateInterval == Azure::DateTime::duration::zero()
+        ? std::chrono::seconds(10)
+        : options.UpdateInterval;
+
+    m_consumerClientDetails = m_consumerClient->GetDetails();
+    m_loadBalancer = std::make_shared<_private::ProcessorLoadBalancer>(
+        m_checkpointStore,
+        m_consumerClientDetails,
+        options.LoadBalancingStrategy,
+        options.PartitionExpirationDuration == Azure::DateTime::duration::zero()
+            ? std::chrono::minutes(1)
+            : std::chrono::duration_cast<std::chrono::minutes>(
+                options.PartitionExpirationDuration));
+  }
+
+  Processor::~Processor()
+  {
+    Log::Stream(Logger::Level::Verbose) << "~Processor.";
+    Stop();
+    m_consumerClient.reset();
+  }
+
+  void Processor::Start(Azure::Core::Context const& context)
+  {
+    m_processorThread = std::thread([this, context]() {
+      try
+      {
+
+        this->RunInternal(context, false);
+      }
+      catch (std::exception& ex)
+      {
+        Log::Stream(Logger::Level::Error) << "Exception caught running processor: " << ex.what();
+      }
+    });
+    m_isRunning = true;
+  }
+
+  // Stop the running processor, waiting for the processor to terminate.
+  void Processor::Stop()
+  {
+    Log::Stream(Logger::Level::Verbose) << "Stop processor.";
```

Comment:
Just another example - as a customer I would see `Stop processor.` in the logs and would have no idea what are we talking about, which processor.

Reply:
As I said earlier: I have an item on my plate to figure out a structured logging story that is (a) easy to use and (b) allows customers to figure out stuff like this and (c) enables ingestion to Azure logging entities.

This is exactly the kind of problem I'm trying to solve. But this pull request is not the place to implement such a solution.